src/output/embedding_output.rs

```diff
@@ -0,0 +1,67 @@
+use ndarray::Array2;
+
+/// [`SingleBatchOutput`] contains the output of a single batch of inference.
+///
+/// In the future, each batch will need to deal with its own post-processing, such as
+/// pooling etc. This struct should contain all the necessary information for the
+/// post-processing to be performed.
+pub struct SingleBatchOutput<'r, 's> {
+    pub session_outputs: ort::SessionOutputs<'r, 's>,
+    pub attention_mask: Array2<i64>,
+}
+
+/// Container struct with all the outputs from the embedding layer.
+///
+/// This will contain one [`SingleBatchOutput`] object per batch/inference call.
+pub struct EmbeddingOutput<'r, 's> {
+    batches: Vec<SingleBatchOutput<'r, 's>>,
+}
+
+impl<'r, 's> EmbeddingOutput<'r, 's> {
+    /// Create a new [`EmbeddingOutput`] from a [`ort::SessionOutputs`] object.
+    pub fn new(batches: impl IntoIterator<Item = SingleBatchOutput<'r, 's>>) -> Self {
+        Self {
+            batches: batches.into_iter().collect(),
+        }
+    }
+
+    /// Consume this [`EmbeddingOutput`] and return the raw session outputs.
+    ///
+    /// This allows the user to perform their custom extractions outside of this
+    /// library.
+    pub fn into_raw(self) -> Vec<SingleBatchOutput<'r, 's>> {
+        self.batches
+    }
+
+    /// Export the output using the given output transformer.
+    ///
+    /// The transformer shall be responsible for:
+    /// - Selecting the output from the session outputs based on the precedence order,
+    /// - Extracting the tensor from the output, then
+    /// - Transform the tensor into the desired output.
+    ///
+    /// The transformer function should take a slice of [`SingleBatchOutput`], and return
+    /// the desired output type.
+    ///
+    /// If any of the steps fail, this function will return an error, including
+    /// the session output not containing the expected precedence keys.
+    pub fn export_with_transformer<R>(
+        &self,
+        // TODO: Convert this to a trait alias when it's stabilized.
+        // https://github.com/rust-lang/rust/issues/41517
+        transformer: impl Fn(&[SingleBatchOutput]) -> anyhow::Result<R>,
+    ) -> anyhow::Result<R> {
+        transformer(&self.batches)
+    }
```

Comment:
Hey. Where is this being used?

Reply:
Currently its not used, more of a public API than anything. If you want to use the same `EmbeddingOutput` twice, then `consume_with_transformer` is not going to work for you.

I wouldn't mind keeping only one of the two really, and let the use do their own memory management.

Reply:
> I wouldn't mind keeping only one of the two really, and let the use do their own memory management.

+1